persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptor.java

```diff
@@ -187,34 +190,40 @@ public void register(EventListener listener) {
     wrapped.ifPresent(w -> w.register(listener));
   }
 
-  private void resolvePeers() {
+  @SuppressWarnings("NonAtomicOperationOnVolatileField")
+  private void resolveAllPeers() {
     if (!Strings.isNullOrEmpty(proxyDnsName)) {
       try {
-        Set<InetAddress> resolved = resolver.resolve(proxyDnsName);
+        externalPeers = resolvePeers(proxyDnsName, externalPeers);
+        internalPeers = resolvePeers(internalProxyDnsName, internalPeers);
+      } catch (UnknownHostException e) {
+        throw new ServerError("Unable to resolve DNS for proxy protocol peers table", e);
+      }
+      scheduler.schedule(this::resolveAllPeers, resolveDelaySecs, TimeUnit.SECONDS);
+    }
+  }
 
-        if (!peers.equals(resolved)) {
-          // Generate listener events based on the differences between this and the previous
-          // resolved peers.
-          Sets.SetView<InetAddress> added = Sets.difference(resolved, peers);
-          Sets.SetView<InetAddress> removed = Sets.difference(peers, resolved);
+  private Set<InetAddress> resolvePeers(String dnsName, Set<InetAddress> current)
+      throws UnknownHostException {
+    Set<InetAddress> resolved = resolver.resolve(dnsName);
+    if (!current.equals(resolved)) {
+      // Generate listener events based on the differences between this and the previous
+      // resolved peers.
+      Sets.SetView<InetAddress> added = Sets.difference(resolved, current);
+      Sets.SetView<InetAddress> removed = Sets.difference(current, resolved);
 
-          for (EventListener listener : listeners) {
-            for (InetAddress peer : added) {
-              listener.onJoinCluster(peer, proxyPort);
-              listener.onUp(peer, proxyPort);
-            }
-            for (InetAddress peer : removed) {
-              tokensCache.remove(peer);
-              listener.onLeaveCluster(peer, proxyPort);
-            }
-          }
-          peers = resolved;
+      for (EventListener listener : listeners) {
+        for (InetAddress peer : added) {
+          listener.onJoinCluster(peer, proxyPort);
+          listener.onUp(peer, proxyPort);
+        }
+        for (InetAddress peer : removed) {
+          tokensCache.remove(peer);
+          listener.onLeaveCluster(peer, proxyPort);
```

Comment:
We are emitting events for both sets of addresses. I don't think we can do better, we don't know in which context each listener operates.

I believe this was already happening before this change, clients that didn't connect with the PROXY header would still receive events for proxied peers.

In the end it shouldn't matter because drivers discard events if they don't recognize the address.

Reply:
Could you elaborate more when `In the end it shouldn't matter because drivers discard events if they don't recognize the address.` can happen?

Reply:
This interceptor is a global component shared across all connections (see [my other comment](https://github.com/stargate/stargate/pull/1180#discussion_r688140706)).

Unlike `interceptQuery()`, `resolveAllPeers()` is NOT executed in the context of a client query, it's a periodic task that runs in the background. So it must emit events for both private and public clients.

Let's say that node A goes down, its public address is 1.1.1.1 and its private address is 127.28.0.1. We emit DOWN events for both addresses.
A client that is connected to a public address sees public addresses in `system.peers`, so it will process the event for 1.1.1.1, but won't know what to do with 127.28.0.1. For a client connected to a private address it's the opposite.

In the Java driver, an unknown address would be handled [here](https://github.com/datastax/java-driver/blob/4.x/core/src/main/java/com/datastax/oss/driver/internal/core/metadata/MetadataManager.java#L351-L355).

Reply:
I agree it's probably not a big deal. There is a way to fix this now because there's a concept of event filtering in the cql module. We could pass down the destination address and filter the events based on matching that to the `ProxyInfo.attributeKey` channel attribute, but maybe it's not worth it.

Reply:
Is there a way to add summary of above as comments in sources, to explain the logic and assumptions (unknown addresses will be dropped quietly)? Just as context for whoever gets to modify the code next.